### PR TITLE
Improve URI handling in `singularity [run/exec/shell/pull]` commands

### PR DESCRIPTION
**Pull request by bauerm97**
**Description of the Pull Request (PR):**

- Addresses some bugs in handling of URI's as pointed out in #2018 
- Changes the defaults of pull to be more sensible
- Adds testing of uri parsing

(First PR of 2, maybe 3, necessary to completely iron out issues presented in #2018 )

### CODE REVIEW

**Review by dtrudg — CHANGES_REQUESTED** [review]
Getting there 😃 

The SIF created if you don't pull to an explicit filename has no extension - i.e. it ends up called just `ubuntu` not `ubuntu.sif`

Also I think this should mirror the 2.x behaviour of including the tag (if any) in the output filename. I expect to be able to pull different tagged versions of ubuntu without them clashing in filename:

In 2.x 
```
08:23 AM $ singularity pull docker://ubuntu:16.04
<snip>
Done. Container is at: ./ubuntu-16.04.simg
```

In this PR...

```
08:24 AM $ singularity pull docker://ubuntu:16.04
ubuntu
<snip>
INFO:   Build complete!
-rwxr-xr-x  1 dave dave 37040128 Sep 19 08:24  ubuntu
```

**Comment by GodloveD** [comment]
@bauerm97 can you be a bit more specific about which of the issues in #2018 are addressed by this PR so that we can better test?

File: src/pkg/util/uri/uri_test.go [hunk 1]
```diff
@@ -35,10 +35,13 @@ func Test_SplitURI(t *testing.T) {
...
 		ref       string
 	}{
-		{"docker no scope", "docker://ubuntu", "docker", "//ubuntu"},
+		{"docker basic", "docker://ubuntu", "docker", "//ubuntu"},
 		{"docker scoped", "docker://user/image", "docker", "//user/image"},
 		{"dave's magical lolcow", "docker://godlovedc/lolcow", "docker", "//godlovedc/lolcow"},
-		{"docker w/ tags", "docker://godlovedc/lolcow:latest", "docker", "//godlovedc/lolcow:latest"},
+		{"docker with tags", "docker://godlovedc/lolcow:latest", "docker", "//godlovedc/lolcow:latest"},
+		{"library basic", "library://image", "library", "//image"},
+		{"library scoped", "library://collection/image", "library", "//collection/image"},
```
Comment (dtrudg) [inline src/pkg/util/uri/uri_test.go]: Technically library can be:

`library://container` or `library://container:tag` (using default entity/collection)
`library://collection/container` or `library://collection/container:tag` (using default entity)
`library://entity/collection/container` or `library://entity/collection/container:tag` (full path)

File: src/pkg/util/uri/uri_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		name      string
+		uri       string
+		transport string
+		ref       string
+	}{
+		{"docker basic", "docker://ubuntu", "docker", "//ubuntu"},
+		{"docker scoped", "docker://user/image", "docker", "//user/image"},
+		{"dave's magical lolcow", "docker://godlovedc/lolcow", "docker", "//godlovedc/lolcow"},
+		{"docker with tags", "docker://godlovedc/lolcow:latest", "docker", "//godlovedc/lolcow:latest"},
+		{"library basic", "library://image", "library", "//image"},
```
Comment (dtrudg) [inline src/pkg/util/uri/uri_test.go:42]: Technically library can be:

`library://container` or `library://container:tag` (using default entity/collection)
`library://collection/container` or `library://collection/container:tag` (using default entity)
`library://entity/collection/container` or `library://entity/collection/container:tag` (full path)

**Review by dtrudg — APPROVED** [review]
:+1: just a note on the library test cases

**Comment by bauerm97** [comment]
> @bauerm97 can you be a bit more specific about which of the issues in #2018 are addressed by this PR so that we can better test?

I actually mispoke, this fixes #2018

**Pull request merged**